### PR TITLE
Enforce singleton pattern to ensure cache persists

### PR DESCRIPTION
**Pull request by quantumew**
See #i-360-content-190109 for full conversation. Enforcing a singleton pattern will help consumers avoid reseting key cache.

### CODE REVIEW

**Review by tylerodonnell — COMMENTED** [review]
Ruby noob, but this is looking good! Let me figure out how to get a branch into my gemfile and I'll let you know if this works

File: lib/authentic/validator.rb [hunk 1]
```diff
@@ -11,7 +11,7 @@ module Authentic
   #
   # Returns boolean.
   def self.valid?(token)
-    Validator.new.valid?(token)
+    Validator.instance.valid?(token)
   end
```
Comment (kdstew) [inline lib/authentic/validator.rb:18]: What would you think about adding a configure method here?
```
def self.configure(opts)
  Validator.instance.configure(opts)
end
```

That way the user doesn't need to deal with or know about the fact that Validator is a singleton and can just run Authentic's setup like `Authentic.configure(opts)`.

File: README.md [hunk 2]
```diff
@@ -37,7 +39,8 @@ Authentic.ensure_valid(request.cookies[:token])
 
 # Manually pass in iss_whitelist
 opts = { iss_whitelist: ['https://articulate.auth0.com/'] }
-validator = Authentic::Validator.new opts
+Authentic::Validator.configure(opts)
```
Comment (kdstew) [inline README.md:42]: I think this needs to be `Authentic::Validator.instance.configure(opts)`

File: lib/authentic/validator.rb [hunk 3]
```diff
@@ -21,20 +21,28 @@ def self.valid?(token)
   #
   # Returns nothing.
   def self.ensure_valid(token)
-    Validator.new.ensure_valid(token)
+    Validator.instance.ensure_valid(token)
   end
 
   # Public: validates JWTs against JWKs.
   class Validator
+    include Singleton
```
Comment (kdstew) [inline lib/authentic/validator.rb]: Another way to do this without using Singleton would be to store `manager` and `iss_whitelist` as class variables that get set with a class method `configure`.  You would then need to throw an error in the initialize method if a manager and iss_whitelist were not set.

Usage would look like:
```
Authentic::Validator.configure(opts)
Authentic::Validator.new.valid?(token)
```

**Review by kdstew — COMMENTED** [review]
Thanks for putting this PR together!  A few comments/suggestions mainly around abstracting away the Singleton or an option to implement without singleton.

**Comment by kdstew** [comment]
@quantumew Thanks for making these

**Comment by kdstew** [comment]
Sorry that was an accidental Close :grimacing:

**Comment by quantumew** [comment]
@kdstew No worries 👍 great suggestions! Will get to them shortly

**Comment by quantumew** [comment]
@kdstew is this what you had in mind? 3ba3549982ffdfc63178fbe7f7b86217b13c768d

**Comment by kdstew** [comment]
@quantumew that looks good!

File: lib/authentic/validator.rb [hunk 4]
```diff
@@ -8,33 +8,49 @@ module Authentic
   # Public: validate JWTs against JWKs using iss whitelist in an environment variable.
   #
   # token - raw JWT.
+  # opts  - Optionally pass configuration options.
   #
   # Returns boolean.
-  def self.valid?(token)
+  def self.valid?(token, opts = {})
+    Validator.configure(opts)
```
Comment (kdstew) [inline lib/authentic/validator.rb]: I think when `opts` is not provided and it defaults to `{}` if `Validator.configure()` had been called previously with specific values those values would be overridden by the defaults in `configure`.  Maybe `Validator.configure` should only be called when `opts` is provided?

Comment (quantumew) [reply lib/authentic/validator.rb]: I was using that to set defaults but you are right, that is not a great approach a722b995ea93d273662f14051be986c598a119c9

File: README.md [hunk 5]
```diff
@@ -37,7 +39,8 @@ Authentic.ensure_valid(request.cookies[:token])
...
+validator = Authentic::Validator.instance
```
Comment (kdstew) [inline README.md]: We need to update to `Authentic::Validator.new` to match recent changes

**Review by kdstew — APPROVED** [review]
Looking good!  I think this is ready to go once the Readme is updated.

File: README.md (hunk 5 above)
Comment (quantumew) [reply README.md]: 48e528105f85fc5745fed699240d94f343590b67

**Pull request merged**